pkg/checker/functools.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2017-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package checker
+
+import (
+	"fmt"
+	"reflect"
+)
+
+func makeCheckRunner(f interface{}, args ...interface{}) (CheckRunner, error) {
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `CheckRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+func makeRemediationRunner(f interface{}, args ...interface{}) (RemediationRunner, error) {
+	if err := validateFunction(f, args...); err != nil {
+		return nil, fmt.Errorf("invalid function: %v", err)
+	}
+	argValues := make([]reflect.Value, 0, len(args))
+	for _, arg := range args {
+		argValues = append(argValues, reflect.ValueOf(arg))
+	}
+	return func() error {
+		ret := reflect.ValueOf(f).Call(argValues)
+		// len(ret) is assumed to be 1 and of type `error`, because of type
+		// `RemediationRunner`
+		if ret[0].IsNil() {
+			return nil
+		}
+		return ret[0].Interface().(error)
+	}, nil
+}
+
+// validateFunction checks whether `f` is a function, and whether
+// the passed parameters have the correct number and types.
+func validateFunction(f interface{}, args ...interface{}) error {
+	typ := reflect.TypeOf(f)
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+	if kind := typ.Kind(); kind != reflect.Func {
+		return fmt.Errorf("invalid check runner, want %v, got %v", reflect.Func, kind)
+	}
+	// check output parameter type(s). The function must return an error
+	// TODO: assess if we want to also allow (value, error) as acceptable return
+	//       types. This means also changing how values are returned in
+	//       `makeCheckRunner` and `makeRemediationRunner`.
+	switch {
+	case typ.NumOut() == 1:
+		// return type must be error
+		if typ.Out(0) != errorType {
+			return fmt.Errorf("invalid return type, want %v, got %v", errorType, typ.Out(0))
+		}
+	default:
+		return fmt.Errorf("invalid number of returned parameters, want 1, got %d", typ.NumOut())
+	}
+	// check input parameter types
+	if typ.IsVariadic() {
+		if len(args) < typ.NumIn()-1 {
```

Comment:
May be it makes sense to move `// check non-variadic arguments` before this `if` to explain where `-1` comes from.

Reply:
indeed - that comment was originally in the right place, fixing this now